### PR TITLE
feat: multiple galleries on same page support with improved (exit)fullscreen change handling

### PR DESCRIPTION
**Pull request by natterstefan**
## Previous Behaviour

Previously, when one rendered multiple galleries on the same page _all_ reacted when pressing the arrow keys _and all_ galleries thought they were in fullscreen mode when one changed to fullscreen.

## Future Behaviour

Now, only _one_ gallery will react to arrow keys when the gallery **is hovered on desktop or swiped on mobile devices**. Additionally, **only the actual fullscreen-gallery will change it's `isFullscreen` state to true**. All others stay the same.

Both changes need to be kept together to enable full multiple-gallery support.

## How to test

Start the demo app and interact with one of the two galleries. Debug the state with the [react extension for chrome](https://chrome.google.com/webstore/detail/react-developer-tools/fmkadmapgofadopljbjfkapdkoienihi?hl=de) and try the arrow keys by simply hovering and pressing the left- or right-arrow key.

## Related (resolved) Issues

- https://github.com/xiaolin/react-image-gallery/issues/451
- smaller (focused) version of https://github.com/xiaolin/react-image-gallery/pull/446

## Arrow Keys work properly for multiple galleries on the same page

proof (gif): https://d.pr/i/R6OxJL/RoOWebWJDh

## Only one gallery's `isFullscreen` state changes in fullscreen mode

proof (gif): https://d.pr/i/AIlQMM/Fo5CIfub6o

## Other Notes

- `example/app.js` does not comply with the provided ESLint rules
- `.eslintrc.json` is the only file with different spaces rule (4 vs the default 2 in other files)
- I recommend adding [prettier](https://prettier.io/) to the project to enable worry-free auto-formatting with eg. [VS Code](https://code.visualstudio.com/), but also works IDE independent (with eg. [husky](https://www.npmjs.com/package/husky) and [lint-staged](https://github.com/okonet/lint-staged)). Note: I do not know your editor settings, but I had to disable some in VS Code to disable formatting but still the editor applied some formatting after saving the file... ==> Example: https://d.pr/i/JAS2Fs/ene8fioHXr

Would be awesome if you could think about adding prettier or set up some editorconfig. thx!

### CODE REVIEW

File: src/ImageGallery.jsx [hunk 1]
```diff
@@ -894,12 +934,12 @@ export default class ImageGallery extends React.Component {
 
     switch (key) {
       case LEFT_ARROW:
-        if (this.canSlideLeft() && !this.intervalId) {
+        if (this.canSlideLeft() && !this.intervalId && isInFocus) {
```
Comment (lumannnn) [inline src/ImageGallery.jsx]: Minor performance tip:
Do the `this.canSlideLeft()` check last as this is the only value which seems to require some calculation.
`isInFocus` and `!this.intervalId` are probably quick to determine and if one evaluates to `false`, other (heavier) checks become obsolete and are not performed.
```suggestion
        if (!this.intervalId && isInFocus && this.canSlideLeft()) {
```

File: src/ImageGallery.jsx [hunk 2]
```diff
@@ -894,12 +934,12 @@ export default class ImageGallery extends React.Component {
...
           this.slideLeft();
         }
         break;
       case RIGHT_ARROW:
-        if (this.canSlideRight() && !this.intervalId) {
+        if (this.canSlideRight() && !this.intervalId && isInFocus) {
```
Comment (lumannnn) [inline src/ImageGallery.jsx]: Same here:

```suggestion
        if (!this.intervalId && isInFocus && this.canSlideRight()) {
```

File: src/ImageGallery.jsx [hunk 3]
```diff
@@ -982,17 +1022,36 @@ export default class ImageGallery extends React.Component {
...
+    const isFullscreen = this.imageGallery.current === fullScreenElement;
+
+    this.setState(
+      {
+        isFullscreen,
+      },
+      () => {
+        if (onScreenChange) {
+          /**
+           * in order to detect the current gallery going changing to
```
Comment (lumannnn) [inline src/ImageGallery.jsx]:
```suggestion
           * in order to detect the current gallery going to change to
```

File: src/ImageGallery.jsx [hunk 4]
```diff
@@ -982,17 +1022,36 @@ export default class ImageGallery extends React.Component {
...
+           * fullscreen, `onScreenChange will provide a second argument.
```
Comment (lumannnn) [inline src/ImageGallery.jsx]:
```suggestion
           * fullscreen, `onScreenChange` will provide a second argument.
```

**Comment by natterstefan** [comment]
Hi @xiaolin, have you seen this already?

**Comment by xiaolin** [comment]
@natterstefan 
Thanks for the PR. I took a look and have a few comments. 

I see that what you're trying to fix is the arrow keys support with multiple galleries. Although this change may fix your issue it also breaks another. Your current change **only** allows the arrow keys to work on the gallery that is moused over. Some developers may not like that as it is very opinionated and works only in a specific way - it's not as intuitive.

The library should be unopinionated for scenarios where multiple galleries or single galleries are present. Adding left and right arrow keys support was an additional complimentary feature and It would be simpler to remove the feature than add logic to restrict them to work only a specific way. 

For issues like #451, your solution in this PR works for you, but there may be other approaches including simply disabling the arrow keys until the viewing gallery is visible on the viewport (this allows the arrow keys to function without hovering over the gallery), but any feature that attempts to fix arrow keys should remain unopinionated and left to the developer to implement to their own product requirement.

Finally, I do agree there is a bug that should be fixed.
> Additionally, only the actual fullscreen-gallery will change it's `isFullscreen` state to true. All others stay the same.

I will gladly accept the PR portion that fixes the specific issue above.

**Comment by natterstefan** [comment]
Hi @xiaolin,

one last try 😅 (then I'll focus on other things again).

> (...) Your current change **only** allows the arrow keys to work on the gallery that is moused over. (...)

This is just _one way_ of solving this _issue_, besides eg. checking if the gallery is visible in the current viewport (as you mentioned it as well, I like it). Thinking more and more about it this would be even a better solution than the one with mouse over. 👍 (But you have no interest in fixing/implementing this)

> (...) Some developers may not like that as it is very opinionated and works only in a specific way - it's not as intuitive. (...)

Not as intuitive? Do you think it's more intuitive when I press the arrow key _anywhere_ on the page and _any_ gallery will react? Really? 🤔

> (...) Adding left and right arrow keys support was an additional complimentary feature and It would be simpler to remove the feature (or set default `disableKeyDown=true`) than add logic to restrict them to work only a specific way. (...)

But it's buggy... I don't believe the way it works right now is a proper feature, taking into account that _all_ galleries react to arrow keys pressed _anywhere_ on the page. This was the reason customers of our company wanted to turn off the arrow key feature entirely, leaving their customers with a bad experience...

> (...) It would be simpler to remove the feature (or set default `disableKeyDown=true`) (...)

moves the burden to developers of the gallery to fix this behaviour by introducing checks which gallery should react when users press an arrow key. And where would be the ideal place to put this check? Is there any event this package provides where I can add this check? 🤔

Because far as I know, I cannot abort sliding. It just happens, because the gallery listens to every arrow key down...

> (...) disabling the arrow keys until the viewing gallery is visible on the viewport (this allows the arrow keys to function without hovering over the gallery) (...)

I like the "visible on the viewport" solution. Would vote for that. 👍

> (...) but any feature that attempts to fix arrow keys should remain unopinionated and left to the developer to implement to their own product requirement.

Okay, fine. I give up. Can you share an example with me/us how one could render multiple galleries properly with arrow key support (sliding only the active gallery)? Looks like you have an solution in your mind without providing a built-in one.

> Finally, I do agree there is a bug that should be fixed.
> > Additionally, only the actual fullscreen-gallery will change it's `isFullscreen` state to true. All others stay the same.
> I will gladly accept the PR portion that fixes the specific issue above.

👏🏻

**Comment by xiaolin** [comment]
> Okay, fine. I give up. Can you share an example with me/us how one could render multiple galleries properly with arrow key support (sliding only the active gallery)? Looks like you have an solution in your mind without providing a built-in one.

It will be similar to your current solution. You basically wrap the gallery with your own component and toggle `disableKeyDown`

e.g.

```
class MyGalleryComponent extends React {
	constructor() {
	  this.state = {
	    disableKeyDown: true
	  }
	}

	enableKeyDown() {
	  this.setState({disableKeyDown: false});
	}
	render () {
		<div onmouseover={this.enableKeyDown}>
		  <ImageGallery disableKeyDown={this.state.disableKeyDown}
		</div>
	}
}



<MyGalleryComponent /> // render multiple custom galleries
<MyGalleryComponent /> // render multiple custom galleries
<MyGalleryComponent /> // render multiple custom galleries
```

**Comment by natterstefan** [comment]
I agree, this could work. In order to finish this PR, you'd like me to remove everything that is connected to:

```jsx
        onMouseOver={this.onFocus}
        onFocus={this.onFocus}
        onMouseOut={this.onBlur}
        onBlur={this.onBlur}
```

https://github.com/xiaolin/react-image-gallery/pull/464/files#diff-eabaad0cc8751d0947757ba5349ae171R1448-R1451

And only leave the rest of the PR?

**Comment by xiaolin** [comment]
I mean to open a new PR and update 

https://github.com/xiaolin/react-image-gallery/blob/d7c79e35c15d6a6e0e7e0f8dc49d70ecac15ecab/src/ImageGallery.jsx#L997

to
```jsx
 this.setState({ isFullscreen: fullScreenElement === this.imageGallery.current });
```

--
> Edit: can't we add the multiple galleries example you provided somewhere in the README? If you don't believe this is "normal" usual use-case we can add it in a "Advanced Examples" section. :D What do you think?

I don't see a reason to document the example as developers should be able to google and implement their own solutions.

**Pull request closed without merging**